core/src/main/java/com/datastax/oss/driver/internal/core/protocol/ByteBufSegmentBuilder.java

```diff
@@ -0,0 +1,174 @@
+/*
+ * Copyright DataStax, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.datastax.oss.driver.internal.core.protocol;
+
+import com.datastax.oss.protocol.internal.Frame;
+import com.datastax.oss.protocol.internal.FrameCodec;
+import com.datastax.oss.protocol.internal.PrimitiveCodec;
+import com.datastax.oss.protocol.internal.Segment;
+import com.datastax.oss.protocol.internal.SegmentBuilder;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import io.netty.buffer.ByteBuf;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelPromise;
+import io.netty.util.concurrent.Future;
+import io.netty.util.concurrent.GenericFutureListener;
+import java.util.ArrayList;
+import java.util.List;
+import net.jcip.annotations.NotThreadSafe;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@NotThreadSafe
+public class ByteBufSegmentBuilder extends SegmentBuilder<ByteBuf, ChannelPromise> {
+
+  private static final Logger LOG = LoggerFactory.getLogger(ByteBufSegmentBuilder.class);
+
+  private final ChannelHandlerContext context;
+  private final String logPrefix;
+
+  public ByteBufSegmentBuilder(
+      @NonNull ChannelHandlerContext context,
+      @NonNull PrimitiveCodec<ByteBuf> primitiveCodec,
+      @NonNull FrameCodec<ByteBuf> frameCodec,
+      @NonNull String logPrefix) {
+    super(primitiveCodec, frameCodec);
+    this.context = context;
+    this.logPrefix = logPrefix;
+  }
+
+  @Override
+  @NonNull
+  protected ChannelPromise mergeStates(@NonNull List<ChannelPromise> framePromises) {
+    // We concatenate multiple frames into one segment. When the segment is written, all the frames
+    // are written.
+    ChannelPromise segmentPromise = context.newPromise();
+    segmentPromise.addListener(
+        future -> {
+          if (future.isSuccess()) {
+            for (ChannelPromise framePromise : framePromises) {
+              framePromise.setSuccess();
```

Comment:
will it wait for all `framePromises`? Could you explain the flow of this merge logic for me?

Reply:
The upper layer (`InFlightHandler` to be specific) is giving us a set of frames. With each frame comes a promise that we'll have to complete to say "I'm done writing".
As it turns out, we decide to process all those frames in a single write. So we craft a single compound promise that will propagate its outcome to all the frame promises. Then we use that promise when writing (`ByteBufSegmentBuilder.processSegment`):
```java
context.write(segment, segmentPromise);
```
When Netty is done writing the segment, it will complete `segmentPromise`, and therefore all the frame promises.

It's the same logic in reverse for large frame: one frame promise becomes multiple segment promises.